Admin/BreadcrumbsBuilder.php

```diff
@@ -0,0 +1,121 @@
+<?php
+
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\AdminBundle\Admin;
+
+use Knp\Menu\ItemInterface as MenuItemInterface;
+
+/**
+ * Stateless breadcrumbs builder (each method needs an Admin object).
+ *
+ * @author Grégoire Paris <postmaster@greg0ire.fr>
+ */
+final class BreadcrumbsBuilder implements BreadcrumbsBuilderInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getBreadcrumbs(AdminInterface $admin, $action)
+    {
+        $breadcrumbs = array();
+        if ($admin->isChild()) {
+            return $admin->getParent()->getBreadcrumbs($action);
+        }
+
+        $menu = $admin->buildBreadcrumbs($action);
+
+        do {
+            $breadcrumbs[] = $menu;
+        } while ($menu = $menu->getParent());
```

Comment:
Overkill IMO. If someone does this, they will do it with code, see that it does not work, and not commit it. Anyway, this is out of scope for this PR, which tries to extract the breadcrumbs builder, not refactor it.